client/src/viya/profile.ts

```diff
@@ -0,0 +1,404 @@
+import { readFileSync, writeFileSync, existsSync } from "fs";
+import { window } from 'vscode';
+
+
+/**
+ * The default compute context that will be used to create a SAS session.
+ */
+const DEFAULT_COMPUTE_CONTEXT = 'SAS Job Execution compute context';
+
+
+/**
+ * Dictionary is a type that maps a generic object with a string key.
+ */
+type Dictionary<T> = {
+  [key: string]: T;
+};
+
+
+/**
+ * ConfigFile is a configuration file manager that supports marshaling
+ * a generic interface.
+ */
+class ConfigFile<T> {
+  protected value: T | undefined;
+
+  constructor(
+    private readonly filename: string,
+    private readonly defaultValue: () => T) {
+    this.getSync();
+  }
+
+  /**
+   * Retreives the configuration {@link T}
+   * 
+   * @param reload {@link Boolean} reloads file before returning {@link T}
+   * @returns Promise<T>
+   */
+  async get(reload = false): Promise<T> {
+    return this.getSync(reload);
+  }
+
+  /**
+   * Synchronous get with optional reload if value is already set 
+   * @param reload {@link Boolean} reloading configuration file
+   * @returns T
+   */
+  getSync(reload = false): T {
+    if (this.value && !reload) {
+      return this.value;
+    }
+
+    // if file exists, parse and set value
+    if (existsSync(this.filename)) {
+      const text = readFileSync(this.filename, 'utf-8');
+      this.value = JSON.parse(text);
+      return this.value;
+    }
+
+    // if file does not exist, set default and 
+    // update, which in turn creates the file.
+    this.updateSync(this.defaultValue());
+    return this.value;
+  }
+
+  /**
+   * Marshal's configuration file based on the T value
+   * @param value 
+   */
+  async update(value: T): Promise<void> {
+    this.updateSync(value);
+  }
+
+  /**
+  * Marshal's configuration file based on the T value
+  * @param value
+  */
+  updateSync(value: T): void {
+    this.value = value;
+    const text = JSON.stringify(this.value, undefined, 2);
+    writeFileSync(this.filename, text);
+  }
+}
+
+/**
+ * Enum that represents the authentication type for a profile.
+ */
+export enum AuthType {
+  TokenFile = 'token-file',
+  Password = 'password',
+  Error = 'error'
+}
+
+/**
+ * Profile is an interface that represents a users profile.  Currently
+ * supports two different authentication flows, token-file and password
+ * flow with the client-id and client-secret.
+ */
+export interface Profile {
+  'sas-endpoint': string;
+  'client-id'?: string;
+  'client-secret'?: string;
+  'compute-context': string;
+  'token-file'?: string;
+  'active'?: boolean;
```

Comment:
Is it possible to store an optional username so that if it exists, I don't have to type each time.

Reply:
Went back and forth on adding the username, since it was not apart of the profile connection we decided to leave out.  We were trying to follow a bit how sas-viya cli creates the configuration file.  They have a separate credentials.json file that stores the username and the refresh token, and the expiration date.  Since we weren't storing the token, we decided to leave the username out.

That being said, this could be an easy add to the current profile object and easy win to reduce a prompt when running the sas code.

Reply:
I have added a `user-name` key to each client-id profile. I figured that only prompting the user with a password was still a bit hard to use, as the user will have to remember which username they set up per-profile. We could revert to the original method, and tell the user what their username is, (e.g. "password for <username>"), however, @jpreece thought of a different idea. We can just pull the username from the profile and use it as a placeholder. This is the direction that I went with it. 
The user now has the opportunity to store a username in the profile, and it will be recalled when they click the "run" command. They then have the option to use the profile username, or change it for a single log on. The only way to update the user-name in the profile is the run the "update profile" command.